web/app/js/components/NetworkGraph.jsx

```diff
@@ -0,0 +1,216 @@
+import _ from 'lodash';
+import ErrorBanner from './ErrorBanner.jsx';
+import PropTypes from 'prop-types';
+import React from 'react';
+import { withContext } from './util/AppContext.jsx';
+import * as d3 from 'd3';
+import 'whatwg-fetch';
+
+
+const defaultSvgWidth = 494;
+const defaultSvgHeight = 295;
+const margin = { top: 0, right: 0, bottom: 10, left: 0 };
+
+const simulation = d3.forceSimulation()
+  .force("link", d3.forceLink().id(function(d) { return d.id; }).distance(100))
+  .force('charge', d3.forceManyBody().strength(-20))
+  .force('center', d3.forceCenter(defaultSvgWidth / 2, defaultSvgHeight / 2));
+
+class NetworkGraph extends React.Component {
+  static defaultProps = {
+    deployments: []
+  }
+
+  static propTypes = {
+    api: PropTypes.shape({
+      cancelCurrentRequests: PropTypes.func.isRequired,
+      fetchMetrics: PropTypes.func.isRequired,
+      getCurrentPromises: PropTypes.func.isRequired,
+      setCurrentRequests: PropTypes.func.isRequired,
+      urlsForResource: PropTypes.func.isRequired,
+    }).isRequired,
+    deployments: PropTypes.arrayOf(PropTypes.object),
+  }
+
+  constructor(props) {
+    super(props);
+    this.api = this.props.api;
+    this.handleApiError = this.handleApiError.bind(this);
+    this.loadFromServer = this.loadFromServer.bind(this);
+    this.updateGraph = this.updateGraph.bind(this);
+
+    this.state = {
+      nodes: [],
+      links: [],
+      pendingRequests: false,
+      error: ""
+    };
+  }
+
+  componentDidMount() {
+    this.svg = d3.select(".network-graph-container")
+      .append("svg")
+      .attr("class", "network-graph")
+      .attr("width", defaultSvgWidth)
+      .attr("height", defaultSvgHeight)
+      .append("g")
+      .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
+
+    this.loadFromServer();
+  }
+
+  componentDidUpdate() {
+    simulation.alpha(1).restart();
+  }
+
+  componentWillUnmount() {
+    this.api.cancelCurrentRequests();
+  }
+
+  loadFromServer() {
+    if (this.state.pendingRequests) {
+      return; // don't make more requests if the ones we sent haven't completed
+    }
+    this.setState({ pendingRequests: true });
+
+    let deployments = _.sortBy(_.map(this.props.deployments, 'name'));
+    let urls = _.map(this.props.deployments, d => {
+      return this.api.fetchMetrics(this.api.urlsForResource("deployment", "") + "&to_name=" + d.name);
+    });
+
+    this.api.setCurrentRequests(urls);
+
+    Promise.all(this.api.getCurrentPromises())
+      .then(results => {
+        let links = [];
+        let nodeList = [];
+
+        _.map(results, (r, i) => {
```

Comment:
I find that sticking to a functional pattern makes javascript more enjoyable to work with. (You'd want to use an `_.each` here purely from a semantics perspective with how you're using it now).

I *think* this does the same thing:

```
const links = _(results)
  .map(r => _.get(r, 'ok.statTables[0].podGroup.rows'))
  .zip(deployments)
  .filter(([rows]) => !_.isEmpty(rows))
  .flatMap(([rows, deployment]) => _.map(rows, row => ({
    source: row.resource.name,
    target: deployment,
  })))
  .value();

const nodes = _(links)
  .flatMap(_.values)
  .uniq()
  .map(n => ({id: n, r: 15}))
  .value();
```